File: validation/differential/differential.ts
// Differential fuzz: Rust privacy-pools helpers vs the 0xbow TS SDK / zk-kit / viem.
//   bun install && bun run differential.ts [opsPerKind]
import { writeFileSync } from "fs";
import { join, resolve } from "path";
import { poseidon } from "maci-crypto/build/ts/hashing.js";
import { LeanIMT } from "@zk-kit/lean-imt";
import { calculateContext } from "@0xbow/privacy-pools-core-sdk";
import { keccak256, encodePacked } from "viem";

const REPO = resolve(import.meta.dir, "../..");
const CASES = join(import.meta.dir, "cases.json");
const P = 21888242871839275222246405745257275088548364400416034343698204186575808495617n;
const N = Number(process.argv[2] ?? 150);
const mod = (x: bigint) => ((x % P) + P) % P;

const rnd = (n: number) => { const b = new Uint8Array(n); crypto.getRandomValues(b); return b; };
const toHex = (b: Uint8Array) => "0x" + Buffer.from(b).toString("hex");
const randField = () => (BigInt(toHex(rnd(32))) % P).toString();
const randAddr = () => toHex(rnd(20));
const randInt = (max: number) => Math.floor(Math.random() * max);

type Op = Record<string, any>;
const ops: Op[] = [];
const ref: string[] = [];
const push = (op: Op, out: string) => { ops.push(op); ref.push(out); };

for (let i = 0; i < N; i++) {
  for (const k of [1, 2, 3]) {
    const inp = Array.from({ length: k }, randField);
    push({ op: "poseidon", in: inp }, poseidon(inp.map(BigInt)).toString());
  }
  const [v, l, n, s] = [randField(), randField(), randField(), randField()];
  push({ op: "nullifierHash", in: [n] }, poseidon([BigInt(n)]).toString());
  push({ op: "precommitment", in: [n, s] }, poseidon([BigInt(n), BigInt(s)]).toString());
  push({ op: "commitment", in: [v, l, n, s] },
    poseidon([BigInt(v), BigInt(l), poseidon([BigInt(n), BigInt(s)])]).toString());

  const pool = randAddr(), asset = randAddr(), chainId = String(randInt(1_000_000) + 1);
  const scope = mod(BigInt(keccak256(encodePacked(["address", "uint256", "address"], [pool as any, BigInt(chainId), asset as any]))));
  push({ op: "scope", pool, asset, chainId }, scope.toString());
  const nonce = String(randInt(1_000_000) + 1);
  const lbl = mod(BigInt(keccak256(encodePacked(["uint256", "uint256"], [scope, BigInt(nonce)]))));
  push({ op: "label", scope: scope.toString(), nonce }, lbl.toString());

  const proc = randAddr();
  const data = toHex(rnd(randInt(4) * 32));
  const ctx = BigInt(calculateContext({ processooor: proc as any, data: data as any }, BigInt(scope) as any)).toString();
  push({ op: "context", processooor: proc, data, scope: scope.toString() }, ctx);

  const m = randInt(20) + 1;
  const leaves = Array.from({ length: m }, randField);
  const t = new LeanIMT<bigint>((a, b) => poseidon([a, b]));
  t.insertMany(leaves.map(BigInt));
  push({ op: "leanRoot", leaves }, t.root.toString());
  const idx = randInt(m);
  const pr = t.generateProof(idx);
  push({ op: "leanProof", leaves, index: idx }, `${pr.index}:${pr.siblings.map((x) => x.toString()).join(",")}`);
}

writeFileSync(CASES, JSON.stringify({ ops }));

const proc = Bun.spawnSync(["cargo", "run", "-q", "--release", "--example", "differential", "--", CASES], { cwd: REPO });
if (proc.exitCode !== 0) { console.error("rust example failed:\n", proc.stderr.toString()); process.exit(1); }
const rust: string[] = JSON.parse(proc.stdout.toString()).out;

let mismatches = 0;
for (let i = 0; i < ops.length; i++) {
  if (rust[i] !== ref[i]) {
    if (++mismatches <= 10) {
      console.error(`MISMATCH [${ops[i].op}]`, JSON.stringify(ops[i]).slice(0, 160));
      console.error("  rust:", rust[i], "\n  ts:  ", ref[i]);
    }
  }
}
const byKind = ops.reduce((m, o) => ((m[o.op] = (m[o.op] || 0) + 1), m), {} as Record<string, number>);
console.log("ops by kind:", byKind);
console.log(`total ops: ${ops.length}, mismatches: ${mismatches}`);
process.exit(mismatches === 0 ? 0 : 1);

File: crates/privacy-pools/examples/differential.rs
//! Differential-testing helper: read a list of ops (JSON), compute each with the
//! Rust helpers, print the results (JSON). Paired with a bun harness that runs
//! the same ops through the 0xbow TypeScript SDK / @zk-kit/lean-imt.
//!
//!   cargo run --release --example differential -- ops.json

use privacy_pools::{
    commitment_hash, context, label, nullifier_hash, poseidon, precommitment, scope, Address,
    Field, LeanImt,
};
use serde_json::Value;

fn f(v: &Value) -> Field {
    Field::from_decimal(v.as_str().unwrap()).unwrap()
}

fn fields(v: &Value) -> Vec<Field> {
    v.as_array().unwrap().iter().map(f).collect()
}

fn addr(v: &Value) -> Address {
    Address::from_hex(v.as_str().unwrap()).unwrap()
}

fn u64v(v: &Value) -> u64 {
    v.as_str().unwrap().parse().unwrap()
}

/// Account-derivation ops, only available with the `account` feature. Returns
/// `None` for non-account ops so the main `match` handles them.
#[cfg(feature = "account")]
fn account_op(op: &Value) -> Option<String> {
    use privacy_pools::Account;
    let pair = |(n, s): (Field, Field)| format!("{},{}", n.to_decimal(), s.to_decimal());
    let mnemonic = || op["mnemonic"].as_str().unwrap();
    let index = || op["index"].as_str().unwrap().parse::<u64>().unwrap();
    Some(match op["op"].as_str().unwrap() {
        "masterKeys" => {
            let k = Account::from_mnemonic(mnemonic()).unwrap().master_keys();
            format!("{},{}", k.nullifier.to_decimal(), k.secret.to_decimal())
        }
        "depositSecret" => pair(
            Account::from_mnemonic(mnemonic())
                .unwrap()
                .deposit_secrets(f(&op["scope"]), index())
                .unwrap(),
        ),
        "withdrawalSecret" => pair(
            Account::from_mnemonic(mnemonic())
                .unwrap()
                .withdrawal_secrets(f(&op["label"]), index())
                .unwrap(),
        ),
        _ => return None,
    })
}

#[cfg(not(feature = "account"))]
fn account_op(_: &Value) -> Option<String> {
    None
}

fn run(op: &Value) -> String {
    if let Some(s) = account_op(op) {
        return s;
    }
    match op["op"].as_str().unwrap() {
        "poseidon" => poseidon(&fields(&op["in"])).unwrap().to_decimal(),
        "nullifierHash" => nullifier_hash(f(&op["in"][0])).unwrap().to_decimal(),
        "precommitment" => precommitment(f(&op["in"][0]), f(&op["in"][1]))
            .unwrap()
            .to_decimal(),
        "commitment" => {
            let a = fields(&op["in"]);
            commitment_hash(a[0], a[1], a[2], a[3])
                .unwrap()
                .to_decimal()
        }
        "scope" => scope(addr(&op["pool"]), u64v(&op["chainId"]), addr(&op["asset"])).to_decimal(),
        "label" => label(f(&op["scope"]), u64v(&op["nonce"])).to_decimal(),
        "context" => {
            let data = hex::decode(op["data"].as_str().unwrap().trim_start_matches("0x")).unwrap();
            context(addr(&op["processooor"]), &data, f(&op["scope"])).to_decimal()
        }
        "leanRoot" => LeanImt::from_leaves(&fields(&op["leaves"]))
            .unwrap()
            .root()
            .unwrap()
            .to_decimal(),
        "leanProof" => {
            let tree = LeanImt::from_leaves(&fields(&op["leaves"])).unwrap();
            let p = tree
                .generate_proof(op["index"].as_u64().unwrap() as usize)
                .unwrap();
            let sibs: Vec<String> = p.siblings.iter().map(Field::to_decimal).collect();
            format!("{}:{}", p.index, sibs.join(","))
        }
        // Root after each successive insert (exercises depth growth / promotion
        // at every size, not just the final tree).
        "leanRootSeq" => {
            let leaves = fields(&op["leaves"]);
            let mut tree = LeanImt::new();
            let mut roots = Vec::with_capacity(leaves.len());
            for leaf in leaves {
                tree.insert(leaf).unwrap();
                roots.push(tree.root().unwrap().to_decimal());
            }
            roots.join(";")
        }
        // Membership proofs for many leaves of one tree (dedups the leaf array).
        "leanProofs" => {
            let tree = LeanImt::from_leaves(&fields(&op["leaves"])).unwrap();
            let proofs: Vec<String> = op["indices"]
                .as_array()
                .unwrap()
                .iter()
                .map(|idx| {
                    let p = tree.generate_proof(idx.as_u64().unwrap() as usize).unwrap();
                    let sibs: Vec<String> = p.siblings.iter().map(Field::to_decimal).collect();
                    format!("{}:{}", p.index, sibs.join(","))
                })
                .collect();
            proofs.join(";")
        }
        other => panic!("unknown op {other}"),
    }
}

fn main() {
    let path = std::env::args()
        .nth(1)
        .expect("usage: differential <ops.json>");
    let j: Value = serde_json::from_slice(&std::fs::read(path).unwrap()).unwrap();
    let out: Vec<String> = j["ops"].as_array().unwrap().iter().map(run).collect();
    println!(
        "{}",
        serde_json::to_string(&serde_json::json!({ "out": out })).unwrap()
    );
}
